lib/tools/android-manifest.js

```diff
@@ -50,40 +50,46 @@ manifestMethods.packageAndLaunchActivityFromManifest = async function packageAnd
  *                 application package.
  */
 manifestMethods.targetSdkVersionFromManifest = async function targetSdkVersionFromManifest (appPath) {
+  log.debug(`Extracting target SDK version of '${appPath}'`);
   const originalAppPath = appPath;
   if (appPath.endsWith(APKS_EXTENSION)) {
     appPath = await this.extractBaseApk(appPath);
   }
 
-  log.info('Extracting target SDK version from the manifest');
-  try {
-    const apkanalyzerPath = await getApkanalyzerForOs(this);
-    const {stdout} = await exec(apkanalyzerPath, ['manifest', 'target-sdk', appPath], {
-      shell: true,
-      cwd: path.dirname(apkanalyzerPath),
-    });
-    if (isNaN(_.trim(stdout))) {
-      throw new Error(`Cannot parse the minimum SDK version from '${stdout}'`);
+  const getTargetSdkViaApkReader = async () => {
+    const apkReader = await ApkReader.open(appPath);
+    const manifest = await apkReader.readManifest();
+    if (manifest.usesSdk && _.isInteger(manifest.usesSdk.targetSdkVersion)) {
+      return manifest.usesSdk.targetSdkVersion;
     }
-    return parseInt(_.trim(stdout), 10);
-  } catch (e) {
-    log.info(`Cannot extract targetSdkVersion using apkanalyzer. Falling back to aapt. ` +
-      `Original error: ${e.message}`);
+    throw new Error('Cannot find the information about targetSdkVersion in the manifest');
+  };
+  const getTargetSdkViaAapt = async () => {
     await this.initAapt();
     const args = ['dump', 'badging', appPath];
-    let output;
-    try {
-      const {stdout} = await exec(this.binaries.aapt, args);
-      output = stdout;
-    } catch (e) {
-      throw new Error(`Fetching targetSdkVersion from '${originalAppPath}' failed. ` +
-        `Original error: ${e.message}`);
-    }
-    const targetSdkVersion = new RegExp(/targetSdkVersion:'([^']+)'/g).exec(output);
+    const {stdout} = await exec(this.binaries.aapt, args);
+    const targetSdkVersion = /targetSdkVersion:'([^']+)'/g.exec(stdout);
     if (!targetSdkVersion) {
-      throw new Error(`targetSdkVersion is not specified in the '${originalAppPath}' application`);
+      log.debug(stdout);
+      throw new Error('Cannot parse the command output');
     }
     return parseInt(targetSdkVersion[1], 10);
+  };
+  const versionGetters = [
+    ['ApkReader', getTargetSdkViaApkReader],
+    ['aapt', getTargetSdkViaAapt],
+  ];
+  for (const [index, [toolName, versionGetter]] of versionGetters.entries()) {
+    try {
+      return await versionGetter();
+    } catch (e) {
+      const msg = `Cannot extract targetSdkVersion of '${originalAppPath}' using ${toolName}. ` +
+        `Original error: ${e.message}`;
+      if (index >= versionGetters.length - 1) {
+        throw new Error(msg);
```

Comment:
I think this error should be outside the `for` loop, and say that neither tool worked.

Reply:
ok will change